enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,244 @@
+
+### Goals
+
+- Allow OLM to automatically recover from failed installations if a new operator is added to a CatalogSource that replaces the last successfully installed operator.
+
+### Non-Goals
+
+- Support Operator rollbacks.
+- Allow cluster admin to perform unsafe/unsane upgrades.
+- Retrying failed installations for operators.
+-- This is supported in limited capacity today as won't immediately fail if it finds that some requirements haven't been met.
+
+## Proposal
+
+OLM allows cluster admins to subscribe to operator updates through the use of subscriptions in a namespace. OLM determines what to install based on the set of subscriptions in a namespace. The list of resources that must be created are then placed within an InstallPlan.
+
+Once an InstallPlan is created, OLM will attempt to create or update each resource listed in the InstallPlan. Operator installs/upgrades generally follow the workflow given below:
+1. A User creates a subscription for a specific operator.
+2. OLM calculates the latest generation of operators that should be installed in a namespace and creates an InstallPlan listing each resource that must exist on cluster.
+3. If the InstallPlan requires manual approval, a User must "approve" it.
+5. OLM creates/updates the manifests defined in the InstallPlan.
+6. If an older CSV exists on the cluster OLM will place it in the "Replacing" phase.
+7. Once successfully applied, OLM marks the new InstallPlan as succeeded.
+8. Once the new CSV reaches the "Succeeded" phase the old CSV is removed from the cluster.
+
+The main reasons that an upgrade may fail include:
+- **Invalid CSV:** The CSV may come in the form of a failing CSV, or in cases of unmet requirements, a forever pending CSV.
+- **Invalid InstallPlan:** Usually occurs because a resource fails to be created or updated. An InstallPlan may fail independently of its CSV. The InstallPlan may fail even with a successfully applied CSV, or even a missing CSV.
+
+So, a failed workflow might look like:
+1. A User creates a subscription for a specific operator.
+2. OLM calculates the latest generation of operators that should be installed in a namespace and creates an InstallPlan listing each resource that must exist on the cluster.
+3. If the InstallPlan requires manual approval, a User must "approve" it.
+5. OLM creates/updates the manifests defined in the InstallPlan
+6. If an older CSV exists on the cluster OLM will place in the "Replacing" phase.
+7a. OLM fails to create/update a resource defined in the InstallPlan and moves the InstallPlan to the "Failed" state.
+7b. OLM installs a CSV and it enters the "Failed" state or forever enters the "Pending" state.
```

Comment:
nit: the ordering here is missing the 4th entry. The markdown looks funny as a result, where it goes from 5 -> 7.

Reply:
Will improve this in the next iteration.